vm-virtio/src/balloon.rs

```diff
@@ -0,0 +1,605 @@
+
+    fn process_queue(&mut self, ev_type: u16) -> result::Result<(), Error> {
+        let queue_index = match ev_type {
+            INFLATE_QUEUE_EVENT => 0,
+            DEFLATE_QUEUE_EVENT => 1,
+            _ => return Err(Error::ProcessQueueWrongEvType(ev_type)),
+        };
+
+        let mut used_desc_heads = [0; QUEUE_SIZE as usize];
+        let mut used_count = 0;
+        let mem = self.mem.memory();
+        for avail_desc in self.queues[queue_index].iter(&mem) {
+            used_desc_heads[used_count] = avail_desc.index;
+            used_count += 1;
+
+            let data_chunk_size = size_of::<u32>();
+
+            // The head contains the request type which MUST be readable.
+            if avail_desc.is_write_only() {
+                error!("The head contains the request type is not right");
+                continue;
+            }
+            if avail_desc.len as usize % data_chunk_size != 0 {
+                error!("the request size {} is not right", avail_desc.len);
+                continue;
+            }
+
+            let mut offset = 0u64;
+            while offset < avail_desc.len as u64 {
+                let addr = avail_desc.addr.checked_add(offset).unwrap();
+                let pfn: u32 = match mem.read_obj(addr) {
+                    Ok(ret) => ret,
+                    Err(e) => {
+                        error!("Fail to read at address 0x{:x}: {:?}", addr.raw_value(), e);
+                        continue;
+                    }
+                };
+                offset += data_chunk_size as u64;
+
+                let gpa = (pfn as u64) << VIRTIO_BALLOON_PFN_SHIFT;
+                if let Ok(hva) = mem.get_host_address(GuestAddress(gpa)) {
+                    let advice = match ev_type {
+                        INFLATE_QUEUE_EVENT => libc::MADV_DONTNEED,
+                        DEFLATE_QUEUE_EVENT => libc::MADV_WILLNEED,
+                        _ => return Err(Error::ProcessQueueWrongEvType(ev_type)),
+                    };
+                    // Need unsafe to do syscall madvise
+                    let res = unsafe {
+                        libc::madvise(
+                            hva as *mut libc::c_void,
+                            (1 << PAGE_SHIFT) as libc::size_t,
+                            advice,
+                        )
+                    };
+                    if res != 0 {
+                        error!("madvise get error {}", io::Error::last_os_error());
+                        continue;
+                    }
+                } else {
+                    error!("Address 0x{:x} is not available", gpa);
+                    continue;
+                }
+            }
+        }
+
+        for &desc_index in &used_desc_heads[..used_count] {
+            self.queues[queue_index].add_used(&mem, desc_index, 0);
+        }
+        if used_count > 0 {
+            self.signal(&VirtioInterruptType::Queue, Some(&self.queues[queue_index]))?;
+        }
+
+        Ok(())
```

Comment:
There are errors in this function that should be propagated up.

Reply:
All added.